google-cloud-pubsub/lib/google/cloud/pubsub/service.rb

```diff
@@ -175,11 +175,18 @@ def list_subscriptions options = {}
         ##
         # Creates a subscription on a given topic for a given subscriber.
         def create_subscription topic, subscription_name, options = {}
-          push_config = if options[:endpoint]
+          push_config = if options[:push_config]
+                          # Handle if push config is provided when creating the subscription
+                          unless options[:push_config].is_a? Google::Cloud::PubSub::Subscription::PushConfig
+                            raise ArgumentError, "push_config is not a valid pubsub PushConfig object"
+                          end
+                          options[:push_config].to_grpc
```

Comment:
Good question. I assumed that the service is a semi-private interface, since it's not well documented, and uses non structured, non documented arguments (ie. keyword arguments in `options`). So in that case throwing an error here feels a bit unorthodox.

Also `options[:attributes]` is not passed forward from `topic.subscribe`.

IMHO, PushConfig construction should be handled upstream (eg. in Topic). Maybe in here `push_config` could even be taken out of the `options` and transformed into an argument on its own - it's quite strange that we're providing constructor parameters of `PushConfig` in the service `options`.

But that's a breaking change. 

PR is already getting quite large. I wonder if we should try to merge something here, and try to address this somewhere else. My suggestion would be that in another PR we would:

- Add a deprecation warning to the options[:endpoint] and options[:attributes] in favour of `PushConfig` object construction in the `Service` class.
- Move the construction of the `PushConfig` upstream to `Topic`, and use `to_grpc` in the `Service` class
- Maybe add `push_auth_email`, `push_auth_audience`, and `push_attributes` to the arguments of `topic.subscribe`

At least we could add an Issue to track this situation.

Reply:
> Also options[:attributes] is not passed forward from topic.subscribe.

Thanks for pointing this out. In that case, it's not a problem.

> I assumed that the service is a semi-private interface, since it's not well documented, and uses non structured, non documented arguments (ie. keyword arguments in options)

Yes, `Service` is private. I think in a separate PR I will convert `options` to separate keyword args, for more clarity.

> This PR is already getting quite large. I wonder if we should try to merge something here, and try to address this somewhere else. 

Agreed! This is great, let's merge it!